File: Unorganized/C/src/lib/Math/src/math.c
#include "math.h"

int fib(int n) {
 if(n < 2) return n;
 else return fib(n-1) + fib(n-2);
}


int fact(int n) {
 if(n == 0) return 1;
 else return n * fact(n-1);
}


int times(int n, int m) {
 return n * m;
}


int double_(int n) {
 return times(n, 2);
}


int induction(int base, int (*comb)(int x, int y), int n) {
 if(n == 0) {
  return base;
 } else {
  return comb(n, induction(base, comb, n-1));
 }
}


int add(int x, int y) {
 return x + y;
}

int mul(int x, int y) {
 return x * y;
}

int sumsqr(int x, int y) {
 return add(mul(x,x),y);
}


int induction_fact(int n) {
 return induction(1, mul, n);
}

int induction_sum_int(int n) {
 return induction(0, add, n);
}

int induction_sum_sqr(int n) {
 return induction(0, sumsqr, n);
}

File: Unorganized/Go/src/Lib/Math/math.go
package math

func Fib(n int) int {
 switch n {
  case 0: return 0
  case 1: return 1
  default: return Fib(n-1) + Fib(n-2)
 }
}

func Fact(n int) int {
 switch n {
  case 0: return 1
  default: return n * Fact(n-1)
 }
}


func Times(n, m int) int {
 return n * m;
}


func Double(n int) int {
 return Times(n, 2)
}


func Induction(base int, comb func(x, y int) int, n int) int {
 if n == 0 {
  return base
 } else {
  return comb(n, Induction(base, comb, (n - 1)));
 } 
}

func Induction_Fact(n int) int {
 mul := func(x, y int) int {
  return x * y
 }
 return Induction(1, mul, n)
}

func Induction_Sum_Int(n int) int {
 add := func(x, y int) int {
  return x + y
 } 
 return Induction(0, add, n)
}

func Induction_Sum_Sqr(n int) int {
 sumsqr := func(x, y int) int {
  return x * x + y
 }
 return Induction(0, sumsqr, n)
}
